Make it easier to detect thread pool exhaustion

RReverser opened the pull request:
Off-the-shelf C++ code doesn't know anything about JS event loop, and "normal" usages of `pthread_create` + `pthread_join` or any form of locks just deadlock silently.

This frequently leads to user confusion and those issues are not the easiest to debug. Some long-term solutions were proposed in https://github.com/emscripten-core/emscripten/issues/9910 and in https://github.com/emscripten-core/emscripten/issues/11554#issuecomment-699082669, but they require more design work.

For now this can at least fix https://github.com/emscripten-core/emscripten/issues/11554 and 1) warn users that the thread pool is exhausted in debug builds (`-s ASSERTIONS=1`) and 2) add an option `PTHREAD_POOL_SIZE_STRICT` to make such exhaustion a hard error in release builds for users who don't expect to create any more threads than preconfigured during the build time.

---

Example output before/after:

```bash
rreverser@rreverser:~/emscripten$ emcc temp.cpp -o temp.js -pthread
rreverser@rreverser:~/emscripten$ node --experimental-wasm-threads --experimental-wasm-bulk-memory temp.js
Before Thread
^C # cancelled because app was locked

rreverser@rreverser:~/emscripten$ ./emcc temp.cpp -o temp.js -pthread
rreverser@rreverser:~/emscripten$ node --experimental-wasm-threads --experimental-wasm-bulk-memory temp.js
Before Thread
Tried to spawn a new thread, but the thread pool is exhausted.
This will result in a deadlock unless the code explicitly breaks out to the event loop.
If you want to increase the pool size, use setting `-s PTHREAD_POOL_SIZE=...`.
If you want to throw an explicit error instead of deadlocking in those cases, use setting `-s PTHREAD_POOL_SIZE_STRICT=1`.
^C # cancelled because app was locked

rreverser@rreverser:~/emscripten$ ./emcc temp.cpp -o temp.js -pthread -s PTHREAD_POOL_SIZE_STRICT=1
rreverser@rreverser:~/emscripten$ node --experimental-wasm-threads --experimental-wasm-bulk-memory temp.js
Before Thread
Tried to spawn a new thread, but the thread pool is exhausted.
This will result in a deadlock unless the code explicitly breaks out to the event loop.
If you want to increase the pool size, use setting `-s PTHREAD_POOL_SIZE=...`.
exception thrown: Error: No available workers in the PThread pool,Error: No available workers in the PThread pool
    at Object.getNewWorker (/usr/local/google/home/rreverser/emscripten/temp.js:2173:17)
    at spawnThread (/usr/local/google/home/rreverser/emscripten/temp.js:2982:28)
    at _pthread_create (/usr/local/google/home/rreverser/emscripten/temp.js:3138:9)
    at <anonymous>:wasm-function[25]:0x923
    at main (<anonymous>:wasm-function[26]:0x956)
    at /usr/local/google/home/rreverser/emscripten/temp.js:1639:22
    at callMain (/usr/local/google/home/rreverser/emscripten/temp.js:3655:15)
    at doRun (/usr/local/google/home/rreverser/emscripten/temp.js:3732:23)
    at run (/usr/local/google/home/rreverser/emscripten/temp.js:3747:5)
    at runCaller (/usr/local/google/home/rreverser/emscripten/temp.js:3633:19)

rreverser@rreverser:~/emscripten$ ./emcc temp.cpp -o temp.js -pthread -s PTHREAD_POOL_SIZE_STRICT=1 -s ASSERTIONS=0
rreverser@rreverser:~/emscripten$ node --experimental-wasm-threads --experimental-wasm-bulk-memory temp.js
Before Thread
exception thrown: Error: No available workers in the PThread pool,Error: No available workers in the PThread pool
    at Object.getNewWorker (/usr/local/google/home/rreverser/emscripten/temp.js:1892:17)
    at spawnThread (/usr/local/google/home/rreverser/emscripten/temp.js:2676:28)
    at _pthread_create (/usr/local/google/home/rreverser/emscripten/temp.js:2832:9)
    at <anonymous>:wasm-function[25]:0x8be
    at main (<anonymous>:wasm-function[26]:0x8f1)
    at Module._main (/usr/local/google/home/rreverser/emscripten/temp.js:2985:60)
    at callMain (/usr/local/google/home/rreverser/emscripten/temp.js:3156:15)
    at doRun (/usr/local/google/home/rreverser/emscripten/temp.js:3222:23)
    at run (/usr/local/google/home/rreverser/emscripten/temp.js:3237:5)
    at runCaller (/usr/local/google/home/rreverser/emscripten/temp.js:3136:19)
# thrown a hard error and exited without long warning
```

---

Another alternative I considered was untying this warning from `ASSERTIONS`. There are two reasons to go that route:

1. This is not always an error and advanced users who return to event loop between thread creation & usage might want to disable this particular warning separately from other assertions.
2. Majority of users might want to see at least a warning even in release builds where other assertions are disabled, since number of threads can depend on the input data.

If that's an acceptable option, then I'd instead define `PTHREAD_POOL_SIZE_STRICT` as follows:

 - `-s PTHREAD_POOL_SIZE_STRICT=0` -> no warnings for users who know what they're doing and adapted their code to break out to the event loop
 - `-s PTHREAD_POOL_SIZE_STRICT=1` -> default option, show warning but still try to create a thread, so that users at least know why there's a deadlock if it occurs
 - `-s PTHREAD_POOL_SIZE_STRICT=2` -> for users who saw the warning and decided to go all-in on the limit, as they are just trying to compile a code to the Web and don't want to mess with event loop APIs

sbc100 commented on src/library_pthread.js:
```diff
@@ -465,6 +465,16 @@ var LibraryPThread = {
+#if !PTHREAD_POOL_SIZE_STRICT
+        err(warning + '\nIf you want to throw an explicit error instead of deadlocking in those cases, use setting `-s PTHREAD_POOL_SIZE_STRICT=1`.');
+#else
+        throw new Error(warning);
```
I wonder if we can detect if the new thread was created on the main thread?

When workers themselves create new threads I believe the can be created and started syncronously (because the its the main thread that does the creation).

Perhaps `getNewWorker` -> `getNewWorker(requestComesFromMainThread)`  ?

RReverser replied:
> When workers themselves create new threads I believe the can be created and started syncronously (because the its the main thread that does the creation).

Maybe, but the detection would be tricky, and I see this warning as a stop-gap solution until we implement one of the proper mentioned fixes that would allow to create new threads at-will.

sbc100 replied:
This warning is only for folks who use the main thread to start other threads and I think it would be to be precise if we can.  
This will limit the number of folks who end up having to disable this warning because they know what they are doing.

At the very least perhaps we can completely disable this new code if PROXY_TO_PTHREAD is used since in that case the pool is not needed.

sbc100 replied:
Also the warning message is wrong for those who start threads from workers.  "This will result in a deadlock unless the code explicitly breaks out to the event loop".. is just not true for workers.. is it?

RReverser replied:
> is just not true for workers.. is it?

Is it always not true or only with PROXY_TO_PTHREAD?

RReverser replied:
Do you perhaps want to add & push those checks to the branch? I have no idea how to implement them and it will be quicker for someone more familiar with codebase.

sbc100 replied:
Sure, I've never done collaborative PR before!   I guess you can just undo anything that I push if you don't want to include it.   Or I could do a PR that targets your PR.. lets see.

RReverser replied:
I think it's only with PROXY_TO_PTHREAD btw. E.g. this still hangs even though the second thread is spawned by a Worker:

```cpp
#include <thread>
#include <iostream>

int main()
{
    std::thread([] {
        std::cout << "1\n";
        std::thread([] {
            std::cout << "2\n";
        }).join();
    }).join();
}
```

sbc100 replied:
> > is just not true for workers.. is it?
> 
> Is it always not true or only with PROXY_TO_PTHREAD?

I think so... but I could be wrong.  

My understanding is that in PROXY_TO_PTHREAD mode no native user code ever runs on the actual browser thread and that things like new thread creation are all proxied back to the browser thread while user's main thread is free to block as long as it wants.    In this mode the browser thread is supposed to be kind of like an invisible background thread that is more of an implementation detail.

Normally when folks run into the problem this warning is about I say either enable PROXY_TO_PTHREAD *or* add POOL_SIZE=N.. either one will fix the problem.

RReverser replied:
So I guess warning should be hidden if and only if (?) `PROXY_TO_PTHREAD` is used? Maybe under `PROXY_TO_WORKER` too?

RReverser replied:
Added the check to avoid warning/error under `PROXY_TO_PTHREAD`.

sbc100 replied:
> I think it's only with PROXY_TO_PTHREAD btw. E.g. this still hangs even though the second thread is spawned by a Worker:
> 
> ```c++
> #include <thread>
> #include <iostream>
> 
> int main()
> {
>     std::thread([] {
>         std::cout << "1\n";
>         std::thread([] {
>             std::cout << "2\n";
>         }).join();
>     }).join();
> }
> ```
> 
> ```shell
> rreverser@rreverser:~/emscripten$ ./emcc temp.cpp -o temp.js -pthread -s PTHREAD_POOL_SIZE=1
> rreverser@rreverser:~/emscripten$ node --experimental-wasm-threads --experimental-wasm-bulk-memory temp.js
> 1
> Tried to spawn a new thread, but the thread pool is exhausted.
> This will result in a deadlock unless the code explicitly breaks out to the event loop.
> If you want to increase the pool size, use setting `-s PTHREAD_POOL_SIZE=...`.
> If you want to throw an explicit error instead of deadlocking in those cases, use setting `-s PTHREAD_POOL_SIZE_STRICT=1`.
> ^C
> ```
> 
> In this case the 2nd warning is still justified and points to a deadlock.



> I think it's only with PROXY_TO_PTHREAD btw. E.g. this still hangs even though the second thread is spawned by a Worker:
> 
> ```c++
> #include <thread>
> #include <iostream>
> 
> int main()
> {
>     std::thread([] {
>         std::cout << "1\n";
>         std::thread([] {
>             std::cout << "2\n";
>         }).join();
>     }).join();
> }
> ```
> 
> ```shell
> rreverser@rreverser:~/emscripten$ ./emcc temp.cpp -o temp.js -pthread -s PTHREAD_POOL_SIZE=1
> rreverser@rreverser:~/emscripten$ node --experimental-wasm-threads --experimental-wasm-bulk-memory temp.js
> 1
> Tried to spawn a new thread, but the thread pool is exhausted.
> This will result in a deadlock unless the code explicitly breaks out to the event loop.
> If you want to increase the pool size, use setting `-s PTHREAD_POOL_SIZE=...`.
> If you want to throw an explicit error instead of deadlocking in those cases, use setting `-s PTHREAD_POOL_SIZE_STRICT=1`.
> ^C
> ```
> 
> In this case the 2nd warning is still justified and points to a deadlock.

Ah I see... in this case the second creation fails because the main thread is blocked so it can't perform the thread creation.

If you remove the outer `join()`  (and move the outer std::thread to global scope) I guess I would work?  (since you are no longer tying up the main thread).

So I guess you are right we can't generally say that worker-originated requests can be serviced ..  :(

sbc100 commented on src/settings.js:
```diff
@@ -1453,6 +1453,17 @@ var USE_PTHREADS = 0;
+// off-the-shelf apps. Those apps without any modifications are most likely
+// to deadlock. This setting ensures that, instead of a deadlock, they get
+// a runtime error instead that can be at least handled from the JS side.
+var PTHREAD_POOL_SIZE_STRICT = 0;
```
I think if we can limit this to requests that come from the main thread we could even make this default to 1?

sbc100 replied:
Then developers who know better can set `PTHREAD_POOL_SIZE_STRICT=0` to avoid the warning

RReverser replied:
> I think if we can limit this to requests that come from the main thread we could even make this default to 1?

Not really - as Alon pointed out to me, it's allowed to create pthreads even on the main thread, you just need to return to the event loop before you lock on any shared vars, and apparently that's how some customers use it.

sbc100 replied:
What I'm saying is that if we can limit this to only those user who start threads on main (and not from workers or PROZY_TO_PTHREAD users) then maybe we can be OK with asking all the remaining valid users of this feature to opt out of this warning/error (since there may be few enough of them).

sbc100 reviewed (commented):
I think this is great in general!

I'm not convinced we need to show this kind of thing in release builds.   Maybe we could compromise and `throw new Error('no workers available');`  in release builds?    (to avoid adding that long warning string)

RReverser commented:
> I'm not convinced we need to show this kind of thing in release builds. Maybe we could compromise and `throw new Error('no workers available');` in release builds? (to avoid adding that long warning string)

Do you mean just to save code size? Yeah sounds reasonable.

RReverser commented:
> > I'm not convinced we need to show this kind of thing in release builds. Maybe we could compromise and `throw new Error('no workers available');` in release builds? (to avoid adding that long warning string)
> 
> Do you mean just to save code size? Yeah sounds reasonable.

Done (but still made error message _a little_ more googlable).

RReverser commented:
@sbc100 I think I addressed your comments; I'd also love to hear your thought on the last part of the PR description - WDYT about detaching `PTHREAD_POOL_SIZE_STRICT=1` into its own setting separately from `ASSERTIONS`?

sbc100 commented on src/library_pthread.js:
```diff
@@ -465,6 +465,20 @@ var LibraryPThread = {
 
     getNewWorker: function() {
       if (PThread.unusedWorkers.length == 0) {
+#if PROXY_TO_PTHREAD
```
This should be `if !PROXY_TO_PTHREAD`, no?

RReverser replied:
Damn :D

RReverser commented:
> I'm not convinced we need to show this kind of thing in release builds.

FWIW the reason I think this is important is that we discovered this only in prod on Squoosh, where apparently codec creates extra threads only for some images but not others (depending on the content). Having an error in this case would help a lot more than users just waiting for encoding to finish and reporting encoder as "slow" when it's in fact deadlocked on that specific input.

sbc100 reviewed (commented):
I like this change a lot now.

Probably should have a test, but otherwise lgtm.

RReverser commented:
Yeah a test would be good. I still wanted to hear thought on the alternative before adding one though.

sbc100 commented:
> @sbc100 I think I addressed your comments; I'd also love to hear your thought on the last part of the PR description - WDYT about detaching `PTHREAD_POOL_SIZE_STRICT=` into its own setting separately from `ASSERTIONS`?

Even though I generally avoid adding new settings, I agree that having this available in production is useful so tying it only to `ASSERTIONS` seems wrong.   Also enabling always in wrong... so it seems a new settings is unavoidable.

Is there are shorter or more descriptive name for this new setting we can come up with?   .. I can't think of one.

RReverser commented:
> Is there are shorter or more descriptive name for this new setting we can come up with? .. I can't think of one.

Me neither, but open to suggestions... Maybe something like `PTHREAD_POOL_ONLY`?

sbc100 commented:
> > Is there are shorter or more descriptive name for this new setting we can come up with? .. I can't think of one.
> 
> Me neither, but open to suggestions... Maybe something like `PTHREAD_POOL_ONLY` / `PTHREAD_ONLY_POOLED`?

`PTHREAD_POOL_ONLY` sounds good.     `PTHREAD_REQUIRE_POOL`?    `PTHREAD_NO_SYNC_CREATE`?

RReverser commented:
`PTHREAD_REQUIRE_POOL` makes it sound like it's enough to have a pool, which it isn't, but the other two sgtm.

RReverser commented:
Meanwhile I tried to work around the Squoosh issue with PROXY_TO_PTHREAD, but ran into more issues: https://github.com/emscripten-core/emscripten/issues/13595 :D

I think that's it for me for today.

kripken commented on src/library_pthread.js:
```diff
@@ -465,6 +465,20 @@ var LibraryPThread = {
+        'This will result in a deadlock unless the code explicitly breaks out to the event loop.\n' +
+        'If you want to increase the pool size, use setting `-s PTHREAD_POOL_SIZE=...`.'
+#if !PTHREAD_POOL_SIZE_STRICT
+        + '\nIf you want to throw an explicit error instead of deadlocking in those cases, use setting `-s PTHREAD_POOL_SIZE_STRICT=1`.'
```
I don't think the text here is precise. There are cases in which a deadlock would not occur, but that option would cause an error to be thrown. Perhaps it could say "instead of the risk of deadlocking"?

RReverser replied:
It says "unless the code explicitly breaks out to the event loop" above - is that not enough?

RReverser replied:
I suppose another case where it won't fail is if threads don't wait for each other, and can finish & be returned to the pool and eventually reused by the "waiting" ones.

I'll reword a bit.

kripken commented on src/settings.js:
```diff
@@ -1453,6 +1453,17 @@ var USE_PTHREADS = 0;
+// the code to adapt it to the Web, and not something that works for
+// off-the-shelf apps. Those apps without any modifications are most likely
+// to deadlock. This setting ensures that, instead of a deadlock, they get
+// a runtime error instead that can be at least handled from the JS side.
```
same comment as above

RReverser replied:
Here it also talks about breaking out of the event loop as a way to avoid deadlock for most part of the comment.

RReverser commented:
Few updates:

 - Switched from `ASSERTIONS` to 3 modes of `PTHREAD_POOL_SIZE_STRICT` as per description & discussion above (0 - no warnings, 1 - warn on pool exhaustion (default), 2 - strict error on trying to create one too many threads).
 - Changed `throw` to instead return EAGAIN from PTHREAD_POOL_SIZE_STRICT check to allow C / C++ to gracefully handle failures.
 - Added two tests for regular pthread_create as well as pthread-in-pthread creation to make sure that both fail gracefully when pool size is too low in strict mode, and pass otherwise.

RReverser commented:
@sbc100 So which name do we settle on in the end? I think that's the only bit left.

sbc100 commented on tests/test_browser.py:
```diff
@@ -3629,6 +3629,24 @@ def test_pthread_c11_threads(self):
+    # Check that it fails instead of deadlocking on insufficient number of threads in the pool.
+    self.btest(path_from_root('tests', 'pthread', 'test_pthread_c11_threads.c'),
+               expected='1',
+               args=['-g2', '-xc', '-std=gnu11', '-pthread', '-s', 'PTHREAD_POOL_SIZE=3', '-s', 'PTHREAD_POOL_SIZE_STRICT=2', '-s', 'TOTAL_MEMORY=64mb'])
```
Is `-xc` needed?    The idea is that it should be automatic based on the filename.  Is that not working for btest?

RReverser replied:
For whatever reason, no. I initially copied those params from the original test, then thought "huh, that's weird, surely -xc is not needed", removed it, it failed to compile, and had to add it back.

sbc100 commented:
I don't have a strong preference for an alternative name.   

lgtm % another round of nits/comments.

RReverser commented:
Looking at CI errors, those seem odd / unrelated? E.g.:

```
AssertionError: JS subprocess failed (/root/.jsvu/v8 --wasm-staging --experimental-wasm-simd test_wasm_intrinsics_simd.js --no-wasm-async-compilation --): 1.  Output:
test_wasm_intrinsics_simd.js:1469: RuntimeError: abort(CompileError: WebAssembly.instantiate(): Compiling function #119:"_Z14i8x16_any_trueDv4_i" failed: local.set[0] expected type i32, found i8x16.popcnt of type s128 @+16838) at Error
    at jsStackTrace (test_wasm_intrinsics_simd.js:1750:19)
    at stackTrace (test_wasm_intrinsics_simd.js:1767:16)
    at abort (test_wasm_intrinsics_simd.js:1463:44)
    at test_wasm_intrinsics_simd.js:1655:7
  var e = new WebAssembly.RuntimeError(what);
          ^
RuntimeError: abort(CompileError: WebAssembly.instantiate(): Compiling function #119:"_Z14i8x16_any_trueDv4_i" failed: local.set[0] expected type i32, found i8x16.popcnt of type s128 @+16838) at Error
    at jsStackTrace (test_wasm_intrinsics_simd.js:1750:19)
    at stackTrace (test_wasm_intrinsics_simd.js:1767:16)
    at abort (test_wasm_intrinsics_simd.js:1463:44)
    at test_wasm_intrinsics_simd.js:1655:7
    at abort (test_wasm_intrinsics_simd.js:1469:11)
    at test_wasm_intrinsics_simd.js:1655:7
```

RReverser commented:
Oh yeah I see main branch is failing too.

sbc100 commented:
Those simd tests are sometimes fragile to v8 and binaryen changes.   @tlively normally knows what is going on.   Certainly not related to this PR.

RReverser commented:
SGTM. Looking forward to comments.

sbc100 reviewed (commented):
Oops.. sorry I forgot to post these comments.

tlively commented:
Merging main should pull in #13610 and fix the SIMD tests.

RReverser commented on tests/pthread/test_pthread_create_pthread.cpp:
```diff
@@ -16,17 +16,19 @@ static void *thread2_start(void *arg)
   EM_ASM(out('thread2_start!'));
   ++result;
-
-  pthread_exit(0);
```
They are, but I find returning a bit clearer unless you intentionally want to break out from the control flow. I can revert this change if you want (but I'd probably still change `0` -> `NULL` for clarity).

RReverser commented on src/settings.js:
```diff
@@ -1453,6 +1453,23 @@ var USE_PTHREADS = 0;
+// the code to adapt it to the Web, and not something that works for
+// off-the-shelf apps. Those apps without any modifications are most likely
+// to deadlock. This setting ensures that, instead of a risking a deadlock,
+// they get a runtime EAGAIN error instead that can be at least gracefully
```
FWIW Github has a suggestion button these days - makes it a bit easier to accept this kind of changes with a button click :)

RReverser replied:
```suggestion
// they get a runtime EAGAIN error instead that can at least be gracefully
```

RReverser commented on src/library_pthread.js:
```diff
@@ -465,11 +465,24 @@ var LibraryPThread = {
+        );
+#endif
+#if PTHREAD_POOL_SIZE_STRICT == 2
+        return;
```
It's just implicitly returning `undefined` like on the line I changed below, as in "not returning any value / Worker". I *think* it's self-describing, but I might be biased and can add a comment if you think it's not.

RReverser commented on src/library_pthread.js:
```diff
@@ -814,13 +832,15 @@ var LibraryPThread = {
-      spawnThread(threadParams);
+      // We don't care about actual exit code, because spawnThread
+      // only returns non-0 as part of the PTHREAD_POOL_SIZE_STRICT check, which
+      // doesn't apply to the PROXY_TO_PTHREAD mode we're in anyway.
```
That's what I originally thought, and added a test to verify, but apparently not - it's taking another branch above:

https://github.com/emscripten-core/emscripten/blob/cdcf0ae5fcf6c9a1d52fa356f2ade821c39aa67e/src/library_pthread.js#L727

sbc100 replied:
If this code is really only designed for `PROXY_TO_PTHREAD` then it should be enclodes in `#ifdef PROXY_TO_PTHREAD` .    It seems like this postMessage version is used whenever there is a transferList (i.e. whenever there is offsetscreen canvas). 

Wouldn't it make more sense to say "only returns non-0 as part of the PTHREAD_POOL_SIZE_STRICT check, which only runs on the main thread and not when ENVIRONMENT_IS_PTHREAD.".

sbc100 replied:
Perhaps it would even better to say something like:  "When using postMessage to spawn a thread we cannot detect failures synchronously so we are forced to assume success and return 0 here".

RReverser replied:
> If this code is really only designed for `PROXY_TO_PTHREAD` then it should be enclodes in `#ifdef PROXY_TO_PTHREAD` . It seems like this postMessage version is used whenever there is a transferList (i.e. whenever there is offsetscreen canvas).

Ah maybe... I'm having hard time understanding what this covers / tries to do tbh, but I suppose you're right that it applies to OffscreenCanvas usages as well.

RReverser replied:
> "When using postMessage to spawn a thread we cannot detect failures synchronously so we are forced to assume success and return 0 here".

FWIW this is not entirely true - we could instead use a shared atomic indicating number of free threads, and synchronously check the value even here. But I don't know if it's worth fixing, especially since I can't yet come up with an example where it'd fail...

sbc100 replied:
But is true today, right?   How about:  "We have no mechanism to detect failure synchronously today so we have to return 0".

RReverser replied:
Reworded & updated.

RReverser commented on src/settings.js:
```diff
@@ -1453,6 +1453,23 @@ var USE_PTHREADS = 0;
 var PTHREAD_POOL_SIZE = '';
 
+// Normally, applications can create new threads even when the pool is empty.
+// When application breaks out to the JS event loop before trying to block on
```
To me, "break out" seems closer to what should happen (since you need to essentially bypass any normal control flow with either Emscripten's event loop APIs, or Asyncify or something else), but I don't feel strongly about it.

RReverser commented on src/library_pthread.js:
```diff
@@ -530,6 +543,10 @@ var LibraryPThread = {
     var worker = PThread.getNewWorker();
 
+    if (!worker) {
+      err('No available workers in the PThread pool');
```
Ah yeah the original message needs to be printed only in ASSERTIONS mode. I'll probably move those messages to a single callsite to avoid confusion.

RReverser replied:
Ok I restored the original logic (short error message) for `PTHREAD_POOL_SIZE_STRICT == 2` case.

RReverser commented on src/settings.js:
```diff
@@ -1453,6 +1453,23 @@ var USE_PTHREADS = 0;
 
+// Normally, applications can create new threads even when the pool is empty.
+// When application breaks out to the JS event loop before trying to block on
+// the thread via `pthread_join` or via manual condvars,
```
```suggestion
// the thread via `pthread_join` or any other blocking primitive,
```

RReverser commented:
@sbc100 Applied suggestions and answered to questions above.

sbc100 reviewed (approved):
lgtm, but maybe you want to wait for https://github.com/emscripten-core/emscripten/pull/13631 to land so you can remove the EM_ASSERT thing?

RReverser commented:
> lgtm, but maybe you want to wait for #13631 to land so you can remove the EM_ASSERT thing?

Sure, sounds good.

juj commented:
How about we create the following set of functions:

```
int emscripten_pthread_available_worker_pool_size();
```
- Returns the number of workers available in the pthread pool.

```
void emscripten_pthread_queue_new_workers(int number, callback);
```
Spins up `number` new pthread workers to the worker pool, calling `callback` `number` times for each worker ready to synchronously spawn a thread.

```
emscripten_pthread_create_sync(...)
```
- Identical to `pthread_create()`, but attempts to synchronously create a thread, and returns ECHILD if no worker resources remain in the pool to do so.

```
emscripten_pthread_create_async(..., callback)
```
- Identical to `pthread_create()`, but thread creation can occur asynchronously, calls `callback` after thread has successfully launched.

Then we can play with a `-s` setting like `-s PTHREAD_CREATE_ASYNC=0/1`, where
 - PTHREAD_CREATE_ASYNC=0: results in `pthread_create()` build time aliasing to `emscripten_pthread_create_sync()`, and
 - PTHREAD_CREATE_ASYNC=1: results in `pthread_create()` build time aliasing to `emscripten_pthread_create_async(..., callback=0)`. (this is what current behavior is)

That way programs will not be locked to a specific behavior of `pthread_create()` for all code that they compile, but if they have a mix of code they did not author and a mix of code they do author, they can adapt their own code to do async creation where desired, and retain the external code they cannot change to have a global default behavior for `pthread_create()`.

RReverser commented:
> How about we create the following set of functions:
> 
> ```
> int emscripten_pthread_available_worker_pool_size();
> ```
> 
> * Returns the number of workers available in the pthread pool.
> 
> ```
> void emscripten_pthread_queue_new_workers(int number, callback);
> ```
> 
> Spins up `number` new pthread workers to the worker pool, calling `callback` `number` times for each worker ready to synchronously spawn a thread.
> 
> ```
> emscripten_pthread_create_sync(...)
> ```
> 
> * Identical to `pthread_create()`, but attempts to synchronously create a thread, and returns ECHILD if no worker resources remain in the pool to do so.
> 
> ```
> emscripten_pthread_create_async(..., callback)
> ```
> 
> * Identical to `pthread_create()`, but thread creation can occur asynchronously, calls `callback` after thread has successfully launched.
> 
> Then we can play with a `-s` setting like `-s PTHREAD_CREATE_ASYNC=0/1`, where
> 
> * PTHREAD_CREATE_ASYNC=0: results in `pthread_create()` build time aliasing to `emscripten_pthread_create_sync()`, and
> * PTHREAD_CREATE_ASYNC=1: results in `pthread_create()` build time aliasing to `emscripten_pthread_create_async(..., callback=0)`. (this is what current behavior is)
> 
> That way programs will not be locked to a specific behavior of `pthread_create()` for all code that they compile, but if they have a mix of code they did not author and a mix of code they do author, they can adapt their own code to do async creation where desired, and retain the external code they cannot change to have a global default behavior for `pthread_create()`.

That sounds useful but probably should be its own feature / issue, rather than in scope of this PR. E.g. it sounds like what we discussed in https://github.com/emscripten-core/emscripten/issues/13595?

Either way, this particular PR is merely about helping to debug and prevent cases where app is deadlocking for otherwise opaque reasons (that are actually caused by something somewhere creating one too many threads).

RReverser commented:
> Returns the number of workers available in the pthread pool.

For this in particular, I was actually thinking that we should repurpose `emscripten_num_logical_cores` to return the preconfigured size in the strict mode, as it seems like a more useful return value for all the cases where someone uses it as "I should create this number of total threads"... but as it can be controversial, also left out of scope of this PR.

sbc100 commented:
#13631 has now landed so you should be able to remove the EM_ASSERT stuff (I think)

RReverser commented:
> #13631 has now landed so you should be able to remove the EM_ASSERT stuff (I think)

Yup, on it right now :)

RReverser commented on tests/pthread/test_pthread_create_pthread.cpp:
```diff
@@ -16,17 +16,19 @@ static void *thread2_start(void *arg)
 static void *thread1_start(void *arg)
 {
-  EM_ASM(out('thread1_start!'));
+  EM_ASM(out('thread1_start!'); debugger);
```
Oops need to remove this too.

RReverser commented:
> > #13631 has now landed so you should be able to remove the EM_ASSERT stuff (I think)
> 
> Yup, on it right now :)

Updated & rebased.

juj commented:
An issue with the way this code landed is that it will now print an error message in the console for all users who `pthread_create()` a thread and there is no worker in the pool to do that. That used to be a perfectly fine use case, but now it is an error..

It is not at all the case that all users would need to synchronously pthread_join() or mutex-lock on a thread, and those users do benefit from a faster application startup time when they do not pause the startup to wait to prepopulate a worker pool; and less memory usage if they do not know what the pool size they need would even be. (Worker pools can take up a lot of memory and time to spin up) Many threading patterns do not require synchronous waits on the created threads (e.g. job pools). But now those users will start getting error messages as well.

In the absence of a greater redesign, `PTHREAD_POOL_SIZE_STRICT == 2` is a nice feature, but `PTHREAD_POOL_SIZE_STRICT == 1` is not that nice because of the above.

Given that this introduced a new -s variable, and if we do want to ultimately go ahead with https://github.com/emscripten-core/emscripten/pull/13591#issuecomment-796949589 , we would be looking to introduce another, and deprecate the first one. That will introduce awkward shuffling on linker vars.

Though maybe that will be unavoidable, I don't unfortunately have a chance from my employer to start working on https://github.com/emscripten-core/emscripten/pull/13591#issuecomment-796949589 either.

RReverser commented:
> that it will now print an error message in the console for all users who `pthread_create()` a thread and there is no worker in the pool to do that. That used to be a perfectly fine use case, but now it is an error..

It can be intentionally disabled via `-s PTHREAD_POOL_SIZE_STRICT=0`. While it's true there are cases where it's possible to safely create threads even when pool is exhausted, on practice those are rare and most beginners and advanced devs alike who are not very familiar with Emscripten get very confused by the deadlocking behaviour with no explanation.

In this case, it's better to show a warning and let advanced users who "know what they're doing" and are sure that their code is safe and not subject to the deadlock, to opt-out from the warning.

RReverser commented:
> Given that this introduced a new -s variable, and if we do want to ultimately go ahead with [#13591 (comment)](https://github.com/emscripten-core/emscripten/pull/13591#issuecomment-796949589) , we would be looking to introduce another, and deprecate the first one.

I think async API is orthogonal; even if it's introduced, the strict behaviour will be still useful. But, as I mentioned above, it would be easier to discuss that API in its own dedicated issue.

sbc100 commented:
I suggest we move any talk of long term solutions back to #11554 or #9910 (I personally like the idea I heard about creating a dedicated worker that is charge of creating pthreads which could (perhaps) allow for the main thread to synchronously start new thread). 

In terms of this specific issue here I guess the question is:  Is it reasonable to ask advanced users to opt into `-s PTHREAD_POOL_SIZE_STRICT=0` to avoid seeing this error/warning?  I tend to agree with @RReverser that this is a worth while trade off.    @juj, do you think there are enough such users and opting out is too much of burden for them?

kripken commented:
I think that we should not show warnings when not in `ASSERTIONS` mode - I though that was the agreed plan but it looks like maybe not? Things were improved in #13646 but I think we should not even show a short warning in a build without assertions enabled.

I 100% agree this is a useful mode, but it is a message for the developer, and should be shown in development builds in an opt-in manner. There should be no other setting than `ASSERTIONS=0` to disable such runtime warnings.

RReverser commented:
> I think that we should not show warnings when not in ASSERTIONS mode - I though that was the agreed plan but it looks like maybe not? 

It was the original plan, but as I said in the description, I think those should be untied from each other, and @sbc100 agreed: https://github.com/emscripten-core/emscripten/pull/13591#issuecomment-790046910.

That is, those warnings are orthogonal to general assertions in that you might want to disable only pthread warning (because you know your code is fine regarding deadlocks) but keep ASSERTIONS enable, or you might want to leave pthread warning for prod builds to make issues more easily discoverable (like the original issue in Squoosh that prompted me making this PR), while disabling the regular ASSERTIONS.


> There should be no other setting than `ASSERTIONS=0` to disable such runtime warnings.

More generally, I think it would be great to have more fine-tuned warning configuration, like any compiler/linter allows you to, but that'd be a bigger project. For now this setting is already different enough (and big enough of a production hazard) to warrant its own config IMO.

RReverser commented:
> Is it reasonable to ask advanced users to opt into `-s PTHREAD_POOL_SIZE_STRICT=0` to avoid seeing this error/warning? I tend to agree with @RReverser that this is a worth while trade off. @juj, do you think there are enough such users and opting out is too much of burden for them?

Just to clarify my reasoning: most of the usecases I've been seeing so far are either 1) creating their own threadpools, spawning tasks onto those and blocking until those are ready or 2) create one-off threads, execute some other code on the main thread, and finally `pthread_join` the side thread to get results. Both of those block in production without any explanation today.

I'm sure there are advanced use-cases where users either spawn & discard threads, letting them run in background and eventually finish, or return to the event pool, but those seem like minority and can still easily opt-out from the warning after upgrade.

The tradeoff between those two categories seems worth it, worst case the latter just get an unnecessary warning that can be immediately disabled, but at least this way we don't have any users for whom Emscripten thread model seems immediately (or, worse, randomly) broken.

kripken commented:
I'm not opposed to fine-grained settings for warnings. That makes sense in general and I'm not opposed to it here.

But I don't think *any* runtime warning messages should be enabled by default. Otherwise users that want a release build with no runtime warnings - the most common use case - need to start accumulating an increasingly-long list of flags to disable them.

RReverser commented:
> need to start accumulating an increasingly-long list of flags to disable them

I understand your reasoning. Personally given the severity of this issue + how common it is, I would go with this option flipped all the way to a strict error by default, with an opt-out for advanced users.

However, that would be a significant breaking change and I thought at least a warning that shows up before a potential deadlock even in release builds would be a suitable middle ground compromise.

After all, it's not breaking anyone and doesn't cost almost anything - the warning is very tiny when `ASSERTIONS=0` and won't affect compressed size nor performance in any noticeable way - yet still provides a helpful message when things go wrong.

However, if you think that's not enough and too disrupting, I guess it's okay to disable it in release builds by default too while keeping settings separate / fine-grained.

juj commented:
> The tradeoff between those two categories seems worth it, worst case the latter just get an unnecessary warning that can be immediately disabled, but at least this way we don't have any users for whom Emscripten thread model seems immediately (or, worse, randomly) broken.

I don't think this error message helps reduce, but rather the opposite, I believe it will increase the number of users who think that the threading model is immediately broken. While previously users could get a deadlock, which I agree is bad, now all users will get an error, which can all the same lead them to think it is broken.

> at least a warning that shows up before a potential deadlock even in release builds would be a suitable middle ground compromise.

Currently the log does not print a warning, but an error (it uses `err()` and not a `console.warn()`).

> Both of those block in production without any explanation today.

I think we should attack the root causes of the block, and not thread creation itself, that can have false positives. How about instead
a) given that we do not have the resources to implement https://github.com/emscripten-core/emscripten/pull/13591#issuecomment-796949589, keep supporting the build mode `PTHREAD_POOL_SIZE_STRICT==2` to enable `pthread_create` to fail if there are no workers in the pool, 
b) if one `pthread_join`s on the main browser thread on a thread that has not been able to start, we should abort execution by printing an error. (`"worker cannot start while main thread is blocked waiting on the thread to finish."`)
c) whenever a new worker is created, a new `pendingWorkerSpawns` counter on the main thread is incremented by one. when that worker actually finishes loading, it sends a message back to main thread to decrement `pendingWorkerSpawns`. Then. if one `pthread_mutex_lock()`s or uses some other blocking wait function on a `pthread_` sync primitive on the main thread, and >5 (>10?) seconds have elapsed and `pendingWorkerSpawns > 0`, we abort the wait and error out with a message `"worker cannot start while main thread is blocked waiting on the thread to operate."` or something like that.

I believe that should detect the practical deadlocks that this issue can cause.

> I personally like the idea I heard about creating a dedicated worker that is charge of creating pthreads

That would change the topology of all created Workers, and require nested Worker support, which has complications and will prevent users from postMessageing anything, we'll need MessagePorts instead. Also even though Chrome finally added support for nested Workers, WebKit does not have that yet: https://bugs.webkit.org/show_bug.cgi?id=22723.

RReverser commented:
> I don't think this error message helps reduce, but rather the opposite, I believe it will increase the number of users who think that the threading model is immediately broken. While previously users could get a deadlock, which I agree is bad, now all users will get an error, which can all the same lead them to think it is broken.

I stongly disagree, good error messages and feedback from tooling only improve development experience over silently failing to work.

Also you're saying "all users" - but it's not all users, the warning is printed only when the pool is exhausted, which affects only 1) beginners who are not yet aware of the `PTHREAD_POOL_SIZE` and `PROXY_TO_PTHREAD` options, and previously would be surprised that their app is immediately stuck after compilation (because `PTHREAD_POOL_SIZE==0`) and 2) users with advanced scenarios who already can avoid the deadlock, and can certainly disable the warning.

It won't affect majority of existing Emscripten projects out there that already work with the Emscripten's threading model, that is where `PTHREAD_POOL_SIZE` is already preconfigured to a reasonable amount or `PROXY_TO_PTHREAD` is used.

> Currently the log does not print a warning, but an error (it uses `err()` and not a `console.warn()`).

I don't mind changing if the output formatting is a concern, this one was just shorter and either way it's just printed to the console and not thrown as an actual error anyway (the app is not aborted in the warning mode, which is what matters for backward compat).

Regarding the other suggestions to improve DX further - please, let's move them to dedicated issues. It's a bit hard to keep track of other discussions on a closed PR :)

kripken commented:
> After all, it's not breaking anyone and doesn't cost almost anything - the warning is very tiny when ASSERTIONS=0 and won't affect compressed size nor performance in any noticeable way - yet still provides a helpful message when things go wrong.

> However, if you think that's not enough and too disrupting, I guess it's okay to disable it in release builds by default too while keeping settings separate / fine-grained.

I get your reasoning, but I do feel strongly here. We have lots of other important warnings, and we don't want to show any of them in a release build by default. It's very surprising for a developer to open the web console and see a developer warning there in a release build.

kripken commented:
ping @RReverser - this will block our next release, so it would be good to resolve. The only urgent thing is to stop showing the new assertion by default.

If you strongly disagree let's keep discussing.

If you are ok with that change but don't have time to make a PR I can make one.

RReverser commented:
> If you are ok with that change but don't have time to make a PR I can make one.

Yeah as I said above I'm okay with that given your reasoning, but couldn't get around to it - got a bit swamped in the last few days. If you can make it, that would be great, thanks!
